File: 2022/july/13-07-2022/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode):
        result = []

        def helper(abc, level):
            if abc is None:
                return
            if level >= len(result):
                result.append([])
            result[level].append(abc.val)

            helper(abc.left, level+1)
            helper(abc.right, level+1)

        helper(root, 0)
        return result


tree = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
print(Solution().levelOrder(tree))

File: 2022/july/13-07-2022/binary-tree-level-order-traversal.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut que = Vec::new();
        let mut ret = Vec::new();
        if let Some(root) = root {
            que.push(root);
        }

        while !que.is_empty() {
            let mut level = Vec::with_capacity(que.len());
            let mut new_que = Vec::with_capacity(que.len() * 2);

            for node in que.iter() {
                level.push(node.borrow().val);
                if let Some(left) = node.borrow_mut().left.take() {
                    new_que.push(left);
                }
                if let Some(right) = node.borrow_mut().right.take() {
                    new_que.push(right);
                }
            }
            ret.push(level);
            que = new_que;
        }
        ret
    }
}